File: Solutions/121_Best_Time_to_Buy_and_Sell_Stock/121_Best_Time_to_Buy_and_Sell_Stock.cs
//Third fastest
public class Solution {
    public int MaxProfit(int[] prices) {
        int maxProfit = 0;
        int minPrice = int.MaxValue;
        
        for(int i = 0; i < prices.Length; i++){
            minPrice = Math.Min(minPrice, prices[i]);
            maxProfit = Math.Max(maxProfit, prices[i] - minPrice);
        }
        
        return maxProfit;
    }
}

//Second fastest
public class Solution {
    public int MaxProfit(int[] prices) {
       int t_i10 = 0, t_i11 = int.MinValue;
        for(int i=0;i< prices.Length;i++){
            t_i10 = Math.Max(t_i10,t_i11 +prices[i]);
            t_i11 =Math.Max(t_i11, -prices[i]);
        }
        return t_i10;
    }

//Fastest
public class Solution {
    public int MaxProfit(int[] prices) {
        if(prices == null || prices.Length == 0)
            return 0;
        int profit = 0;
        int buy =  Int32.MaxValue;
        for(int i= 0; i < prices.Length; i++){
            if(prices[i] < buy){
                buy = prices[i];
            }
            else if(prices[i] - buy> profit){
                profit = prices[i] - buy;
            }
        }
        return profit;
    }
}

File: Solutions/121_Best_Time_to_Buy_and_Sell_Stock/121_Best_Time_to_Buy_and_Sell_Stock.js
var maxProfit = function(prices) {
    let dif = 0
    let min = Infinity
    for(let price of prices){
        min = Math.min(price, min)
        if(min < price) dif = Math.max(dif, (price - min))
    }
    return dif
};

//Fastest
/**
 * @param {number[]} prices
 * @return {number}
 */

var maxProfit = function(prices) {
    let buyPrice = Infinity, maxProfit = 0;
    for (let i = 0; i < prices.length - 1; i++) {
        if (prices[i + 1] > prices[i]) {
            buyPrice = Math.min(buyPrice, prices[i]);
            maxProfit = Math.max((prices[i+1] - buyPrice), maxProfit)
        } 
    }
    return maxProfit
};